src/elf.rs

```diff
@@ -370,27 +370,253 @@ impl<C: ContextObject> Executable<C> {
+                _ => SBPFVersion::Reserved,
+            }
+        };
+        if !config.enabled_sbpf_versions.contains(&sbpf_version) {
+            return Err(ElfError::UnsupportedSBPFVersion);
+        }
+
+        let mut executable = if sbpf_version.enable_stricter_elf_headers() {
+            Self::load_with_strict_parser(bytes, loader)?
         } else {
-            aligned = AlignedMemory::<{ HOST_ALIGN }>::from_slice(bytes);
-            aligned.as_slice()
+            Self::load_with_lenient_parser(bytes, loader)?
         };
-        Self::load_with_parser(&Elf64::parse(bytes)?, bytes, loader)
+        executable.sbpf_version = sbpf_version;
+        Ok(executable)
     }
 
-    fn load_with_parser(
-        elf: &Elf64,
+    /// Loads an ELF without relocation
+    fn load_with_strict_parser(
+        bytes: &[u8],
+        loader: Arc<BuiltinProgram<C>>,
+    ) -> Result<Self, ElfParserError> {
+        use crate::elf_parser::{
+            consts::{
+                ELFMAG, EV_CURRENT, PF_R, PF_W, PF_X, PT_GNU_STACK, PT_LOAD, SHN_UNDEF, STT_FUNC,
+            },
+            types::{Elf64Ehdr, Elf64Shdr, Elf64Sym},
+        };
+
+        let aligned_memory = AlignedMemory::<{ HOST_ALIGN }>::from_slice(bytes);
+        let elf_bytes = aligned_memory.as_slice();
+
+        let (file_header_range, file_header) = Elf64::parse_file_header(elf_bytes)?;
+        let program_header_table_range = mem::size_of::<Elf64Ehdr>()
+            ..mem::size_of::<Elf64Phdr>()
+                .saturating_mul(file_header.e_phnum as usize)
+                .saturating_add(mem::size_of::<Elf64Ehdr>());
+        if file_header.e_ident.ei_mag != ELFMAG
+            || file_header.e_ident.ei_class != ELFCLASS64
+            || file_header.e_ident.ei_data != ELFDATA2LSB
+            || file_header.e_ident.ei_version != EV_CURRENT as u8
+            || file_header.e_ident.ei_osabi != ELFOSABI_NONE
+            || file_header.e_ident.ei_abiversion != 0x00
+            || file_header.e_ident.ei_pad != [0x00; 7]
+            || file_header.e_type != ET_DYN
+            || file_header.e_machine != EM_SBPF
+            || file_header.e_version != EV_CURRENT
+            // file_header.e_entry
+            || file_header.e_phoff != mem::size_of::<Elf64Ehdr>() as u64
+            // file_header.e_shoff
+            // file_header.e_flags
+            || file_header.e_ehsize != mem::size_of::<Elf64Ehdr>() as u16
+            || file_header.e_phentsize != mem::size_of::<Elf64Phdr>() as u16
+            || file_header.e_phnum < EXPECTED_PROGRAM_HEADERS.len() as u16
+            || program_header_table_range.end >= elf_bytes.len()
+            || file_header.e_shentsize != mem::size_of::<Elf64Shdr>() as u16
+            // file_header.e_shnum
+            || file_header.e_shstrndx >= file_header.e_shnum
+        {
+            return Err(ElfParserError::InvalidFileHeader);
+        }
+
+        const EXPECTED_PROGRAM_HEADERS: [(u32, u32, u64); 5] = [
```

Comment:
The `R` `W` `X` flags here don't match with [SIMD-0189](https://github.com/solana-foundation/solana-improvement-documents/pull/189).